pkg/prober/prober_windows.go

```diff
@@ -0,0 +1,104 @@
+	root, err := syscall.UTF16PtrFromString("Root")
+	if err != nil {
+		return nil, fmt.Errorf("[GetSystemCertPoolWindows] unable to return UTF16 pointer: %v", syscall.GetLastError())
+	}
+	if root == nil {
+		return nil, fmt.Errorf("[GetSystemCertPoolWindows] UTF16 pointer for Root returned nil: %v", syscall.GetLastError())
+
+	}
+	// win32 reference: https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certopensystemstorea
+	// If the function succeeds, it returns a handle to the specified certificate store.
+	storeHandle, err := syscall.CertOpenSystemStore(0, root)
+	if err != nil {
+		return nil, fmt.Errorf("[GetSystemCertPoolWindows] unable to open system certContext store: %v", syscall.GetLastError())
+	}
+
+	var certs []*x509.Certificate
+	var certContext *syscall.CertContext
+
+	// ref for why flags value is 0: https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certclosestore
+	defer func(store syscall.Handle, flags uint32) {
+		_ = syscall.CertCloseStore(store, flags)
+	}(certContext.Store, 0)
+
+	for {
+		// the for loop + the syscall for `CertEnumCertificatesInStore` will iterate through
+		// all available certContexts in the specified certificate store and
+		// return a single certContext containing the next certificate in the store
+		certContext, err = syscall.CertEnumCertificatesInStore(storeHandle, certContext)
+		if err != nil {
+			if errno, ok := err.(syscall.Errno); ok {
+				if errno == CRYPT_E_NOT_FOUND {
+					// if the error returned here is CRYPTO_E_NOT_FOUND, that indicates no certificates were found.
+					// This happens if the store is empty or if the function reached the end of the store's list.
+					// ref: https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certenumcertificatesinstore#return-value
+					logrus.Debugf("[GetSystemCertPoolWindows] no certificates were returned from the root CA store for probe (%s)", probeName)
+					break
+				}
+			}
+			logrus.Errorf("[GetSystemCertPoolWindows] unable to enumerate certs in system certContext store for probe (%s): %v", probeName, syscall.GetLastError())
+		}
+		if certContext == nil {
+			logrus.Errorf("[GetSystemCertPoolWindows] certificate context returned from syscall is nil for probe (%s)", probeName)
+			break
+		}
+
+		// buf is a ~1048 kilobyte array that serves as a buffer holding the encoded value
+		// for each CA certificate in the Windows root CA store equal to the length of the certContext pointer
+		// which contains a certificate from the Root CA store
+		// we use a binary shift to create a ~1048 Kb buffer (slightly larger than 1 megabyte)
+		// [1 << 20]byte -> (1*2)^20 = 1048576 bytes
+		// maximum size of a Windows certificate store is 16 kilobytes and is not related to number of certificates
+		// (1048576*8)/4096 = 2048 (amount of 4096-bit certificates that can be stored)
+		buf := (*[1 << 20]byte)(unsafe.Pointer(certContext.EncodedCert))[:certContext.Length]
```

Comment:
One more question - you said up above that
> 		// `CertEnumCertificatesInStore` will iterate through
>		// all available certContexts in the specified certificate store and
>		// return a single certContext containing the next certificate in the store

in that case, the certContext should only ever be 16kb, the max size of a single certificate entry, right? We don't need to over-allocate for the size of the entire store, since that's not what this is going to point at.

Reply:
I actually misinterpreted the syscall and that comment is not a proper description. It only fetches one cert at a time. I will update the comments I included to indicate that.

I am using the workaround referenced here, https://github.com/golang/go/issues/16736#issuecomment-540373689, which does not explain why they chose `(1*2)^20`. I assume it had to do with being divisible by 8 but it does make more sense to use 16kb. I'll get this change added

Reply:
Hmm, I see some similar work in https://github.com/cloudfoundry/systemcerts/blob/master/root_windows.go, and I can't find anything in the MS API docs regarding the max size for the [_CERT_CONTEXT struct](https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_context), so I guess 1024k is a fine worst-case value. You might update the comment to indicate that we're sizing for a single context (certificate) and not the whole store, though. That was a little confusing for me.

Reply:
Since we can't figure out exactly how large it'll be, maybe lets guard against making it too small? Otherwise, if we error on the short size, we'll crash when trying to access past the end of the array.

```suggestion
		// put this up at the top
		const maxEncodedCertLen := 1 << 20
		if certContext.Length > maxEncodedCertLen {
				return fmt.Errorf("invalid CertContext length %d, certContext.Length)
		}
		buf := (*[maxEncodedCertLen]byte)(unsafe.Pointer(certContext.EncodedCert))[:certContext.Length]
```

Reply:
agree w/ guarding against that situation, pushed the changes you suggested